src/line_launcher/mod.rs

```diff
@@ -73,69 +143,98 @@ impl LineLauncher {
                 PlayingState::NotPlaying if beat_message.is_beginning_of_measure() => {
-                    state = PlayingState::Playing {
+                    *state = PlayingState::Playing {
                         line_index: rand::thread_rng().gen_range(0..self.lines.len()),
                         next_note_index: 0,
                         pitch_offset: progression_state.current_chord().pitch.index(),
+                        has_fired_previous_note_off: true,
                     };
-                    self.possibly_trigger_notes(state, &mut midi_message_sender, beat_message)
-                }
-                PlayingState::Playing { .. } => {
-                    self.possibly_trigger_notes(state, &mut midi_message_sender, beat_message)
+                    self.possibly_trigger_notes(
+                        *state,
+                        &midi_message_sender,
+                        beat_message,
+                        &note_off_sender,
+                    )
                 }
-                _ => state,
+                PlayingState::Playing { .. } => self.possibly_trigger_notes(
+                    *state,
+                    &midi_message_sender,
+                    beat_message,
+                    &note_off_sender,
+                ),
+                _ => *state,
             };
         }
     }
 
     fn possibly_trigger_notes(
         &self,
         state: PlayingState,
-        midi_message_sender: &mut MidiMessageSender,
+        midi_message_sender: &Arc<Mutex<MidiMessageSender>>,
         beat_message: BeatNumber,
+        note_off_sender: &Sender<NoteOffInstruction>,
     ) -> PlayingState {
         match state {
             PlayingState::Playing {
                 line_index,
                 next_note_index,
                 pitch_offset,
+                has_fired_previous_note_off,
             } => {
                 let line = &self.lines[line_index];
                 let mut did_trigger_note_off = false;
                 if next_note_index > 0 {
                     let last_played_note = &line.notes[next_note_index - 1];
-                    if beat_message.minus_sixteenths(last_played_note.duration)
-                        == last_played_note.start
+                    if !has_fired_previous_note_off
+                        && beat_message.minus_sixteenths(last_played_note.duration)
+                            == last_played_note.start
                     {
                         midi_message_sender
+                            .lock()
+                            .unwrap()
                             .fire_note_off(last_played_note.note.step(pitch_offset).unwrap());
                         did_trigger_note_off = true;
                     }
                 }
                 if next_note_index == line.notes.len() {
-                    return if did_trigger_note_off {
+                    return if did_trigger_note_off || has_fired_previous_note_off {
```

Comment:
I think this is working as letting `possibly_trigger_notes()` always do the state transition back to `NotPlaying` after the final note off (rather than needing to also teach `NoteOffTriggerer` how to do this state transition when it does the final note off of a line)